cmd/podman/exec.go

```diff
@@ -51,84 +46,34 @@ func init() {
 	flags.IntVar(&execCommand.PreserveFDs, "preserve-fds", 0, "Pass N additional file descriptors to the container")
 	flags.StringVarP(&execCommand.Workdir, "workdir", "w", "", "Working directory inside the container")
 	markFlagHiddenForRemoteClient("latest", flags)
+	markFlagHiddenForRemoteClient("preserve-fds", flags)
 }
 
 func execCmd(c *cliconfig.ExecValues) error {
-	args := c.InputArgs
-	var ctr *libpod.Container
-	var err error
-	argStart := 1
-	if len(args) < 1 && !c.Latest {
-		return errors.Errorf("you must provide one container name or id")
-	}
-	if len(args) < 2 && !c.Latest {
-		return errors.Errorf("you must provide a command to exec")
-	}
-	if c.Latest {
-		argStart = 0
-	}
-	cmd := args[argStart:]
-	runtime, err := libpodruntime.GetRuntime(getContext(), &c.PodmanCommand)
-	if err != nil {
-		return errors.Wrapf(err, "error creating libpod runtime")
-	}
-	defer runtime.Shutdown(false)
-
 	if c.Latest {
-		ctr, err = runtime.GetLatestContainer()
-	} else {
-		ctr, err = runtime.LookupContainer(args[0])
-	}
-	if err != nil {
-		return errors.Wrapf(err, "unable to exec into %s", args[0])
-	}
-
-	if c.PreserveFDs > 0 {
-		entries, err := ioutil.ReadDir("/proc/self/fd")
-		if err != nil {
-			return errors.Wrapf(err, "unable to read /proc/self/fd")
-		}
-		m := make(map[int]bool)
-		for _, e := range entries {
-			i, err := strconv.Atoi(e.Name())
-			if err != nil {
-				if err != nil {
-					return errors.Wrapf(err, "cannot parse %s in /proc/self/fd", e.Name())
-				}
-			}
-			m[i] = true
+		if len(c.InputArgs) < 1 {
+			return errors.Errorf("you must provide a command to exec")
 		}
-		for i := 3; i < 3+c.PreserveFDs; i++ {
-			if _, found := m[i]; !found {
-				return errors.New("invalid --preserve-fds=N specified. Not enough FDs available")
-			}
+	} else {
+		switch {
+		case len(c.InputArgs) < 1:
+			return errors.Errorf("you must provide one container name or id")
+		case len(c.InputArgs) < 2:
+			return errors.Errorf("you must provide a command to exec")
 		}
-
 	}
 
-	// ENVIRONMENT VARIABLES
-	env := map[string]string{}
-
-	if err := parse.ReadKVStrings(env, []string{}, c.Env); err != nil {
-		return errors.Wrapf(err, "unable to process environment variables")
-	}
-	envs := []string{}
-	for k, v := range env {
-		envs = append(envs, fmt.Sprintf("%s=%s", k, v))
+	ctx := getContext()
+	runtime, err := adapter.GetRuntime(ctx, &c.PodmanCommand)
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
 	}
+	defer runtime.Shutdown(false)
 
-	streams := new(libpod.AttachStreams)
-	streams.OutputStream = os.Stdout
-	streams.ErrorStream = os.Stderr
-	streams.InputStream = os.Stdin
-	streams.AttachOutput = true
-	streams.AttachError = true
-	streams.AttachInput = true
-
-	err = ctr.Exec(c.Tty, c.Privileged, envs, cmd, c.User, c.Workdir, streams, c.PreserveFDs)
-	if errors.Cause(err) == libpod.ErrCtrStateInvalid {
-		exitCode = 126
+	exitCode, err = runtime.ContainerExecute(ctx, c)
```

Comment:
Nit: let's rename this to `ContainerExec` - the `Exec` term is pretty standardized now to "start an additional process in an existing container"